### PR TITLE
Move all of our pipelines to use the global-build-job to build CoreCLR, Mono, and Libraries assets

### PR DESCRIPTION
**Pull request by jkoritzinsky**
When we first consolidated dotnet/coreclr, dotnet/corefx, dotnet/core-setup, and mono/mono in .NET 5, we brought in separate product build, test build, and test run pipeline jobs, as before consolidation we were seeing increased build throughput and reliability by having separate jobs, ensuring we built each asset exactly once, and a complex matrix of how they interacted.

Over time, we've removed a few of these jobs and allowed double-building to reduce dependencies:

- Libraries removed their separate "build test" job.
- The CoreCLR/Mono "build test" job now builds the assets it needs directly instead of depending on other jobs.
- The Libraries job now builds System.Private.CoreLib in-job (and now can even just build the ref-assembly for CoreLib)
 
Additionally, we've moved many jobs to build the product (and sometimes test) in one job using the "Global build" template:

- All Mono devices jobs use the "global build" to build the product (and build runtime tests if needed)
- All NativeAOT jobs use the "global build" template
- Nearly all community-supported legs build using the "global build" template.
- A few outerloop pipelines use the "global build" template for simplicity.

Today, our pipelines are split between using the "Global build" template that directly invokes the root build script to build the product in one command and the separate "build job" templates for building specific subsets (CoreCLR, Libraries, Mono, Installer) separately.

This has added significant complexity in many of our YAML scripts to support both cases and confusion about when to use one scenario or the other. For example, some jobs in the "runtime" pipeline use the global build job template to build the product and run tests whereas others don't.

Recently, we moved the official build to use the "global build" template for every job as a) we needed to move in this direction for the VMR and b) we believed that due to job queue time and upload/download time, we'd save on build execution time.

We saw decent improvements, as well as a significant reliability boost by merging these jobs together in the official build.

This PR takes us the next step: Merge all CoreCLR, Mono, and Libraries product build steps (and installer build steps when we aren't running tests) into single jobs instead of being split into 2 or 3 jobs (as all of the Mono runs also required some components of the corresponding CoreCLR build).

In addition, to avoid ending up in a split situation where many of our outerloop legs are still on the older system while PR and official builds are all on the new system, I have moved every pipeline to the "global build" model, not just the PR pipeline.

### CoreCLR + Libraries + Installer (w/o Installer Tests)

```mermaid
flowchart TD

subgraph Old Model
direction LR
CLR[CoreCLR Build Job]
Libs[Libraries Build Job]
Host[Installer Build Job]
CLR --> Host
Libs --> Host
BRT[Build Runtime Tests] --> RT
CLR --> RT[Runtime Tests]
Libs --> RT
CLR --> LT[Libraries Tests]
Libs --> LT
end
subgraph New Model
Global[Global build job: CLR + Libs + Host + Packs]
GBRT[Build Runtime Tests]
GRT[Runtime Tests]
GLT[Libraries Tests]
Global --> GRT
Global --> GLT
GBRT --> GRT
end
```

### CoreCLR + Libraries + Installer w/ Tests

```mermaid
flowchart TD

subgraph Old Model
direction LR
CLR[CoreCLR Build Job]
Libs[Libraries Build Job]
Host[Installer Build and Test]
CLR --> Host
Libs --> Host
BRT[Build Runtime Tests] --> RT
CLR --> RT[Runtime Tests]
Libs --> RT
CLR --> LT[Libraries Tests]
Libs --> LT
end
subgraph New Model
Global[Global build job: CLR + Libs]
GI[Installer Build and Test]
GBRT[Build Runtime Tests]
GRT[Runtime Tests]
GLT[Libraries Tests]
Global --> GRT
Global --> GLT
GBRT --> GRT
Global --> GI
end
```

### Mono Runtime Tests

```mermaid
flowchart TD

subgraph Old Model
direction LR
CLR[CoreCLR Build Job, all of CoreCLR]
Libs[Libraries Build Job]
Mono[Mono Build Job]
BRT[Build Runtime Tests] --> RT
CLR --> RT[Runtime Tests]
Libs --> RT
Mono --> RT
end
subgraph New Model
Global[Global build job: Mono + CoreRun + ilasm/dasm + Libs]
GBRT[Build Runtime Tests]
GRT[Runtime Tests]
Global --> GRT
GBRT --> GRT
end
```

## YAML Template Flow Changes

In addition to the job simplifications above, there's also some YAML template simplification as well. In addition to removing the `build-coreclr-and-libraries-job` template, I consolidated the general flow of the JIT outerloop pipelines for CoreCLR tests into a single pipeline template, simplifying defining them.

Additionally, I moved all outerloop pipelines that run libraries tests to use the global build job with the "Send to Helix" extra step instead of shuffling files between jobs.

## Q&A

- Why do we still have separate jobs per vertical in the PR pipeline?
    - We don't want to wait to run the runtime tests until the libraries tests pass or vice versa. Additionally, building the runtime tests still takes a while (~30 minutes), so we don't want to make that a requirement before any tests start to run until we've reduced that time.
- Why are installer tests still in a separate job?
    - Installer tests are very quick, but we don't want to block libraries and runtime tests from running if installer tests fail?
- Why is installer build still in the installer test job?
    - It's easier given current infrastructure to keep the "Installer Build and Test" job as-is rather than moving the building steps out entirely.

For validation, I'm happy to run any requested pipelines on this PR. I obviously can't run them all, but I'm happy to run a good spread to validate that I didn't break anything (I've been running various pipelines throughout working on this to validate the experience as I went).

### CODE REVIEW

**Comment by agocke** [comment]
This looks great! I'll note that this diff is -1000 lines, so great simplification.

For "CoreCLR + Libraries + Installer (Installer Tests)" -- I don't see an installer step. Is running the installer tests moved somewhere else?

**Comment by jkoritzinsky** [comment]
That's on me, I mistitled that chart. That's for the scenario without installer tests. I'll go correct that.

**Comment by akoeplinger** [comment]
Amazing work!

> > Why are installer tests still in a separate job?
>
> Installer tests are very quick, but we don't want to block libraries and runtime tests from running if installer tests fail?

Right now the installer tests run on the build machine, when we move them to run in Helix (which is tracked by https://github.com/dotnet/runtime/issues/77800) then they'd run concurrently with the other tests and we could remove the separate installer jobs.

File: eng/pipelines/coreclr/jit-experimental.yml
Comment (jakobbotsch) [inline eng/pipelines/coreclr/jit-experimental.yml:24]: Shouldn't it be `jit-experimental`?

**Comment by jakobbotsch** [comment]
Do you mind making a simple change in the JIT source change to make sure the relevant pipelines are triggered there?

**Comment by jakobbotsch** [comment]
jit-cfg seems to be failing: https://dev.azure.com/dnceng-public/public/_build/results?buildId=587562&view=results

File: eng/pipelines/runtime.yml
Comment (jkoritzinsky) [reply eng/pipelines/runtime.yml:1354]: The runs for the runtime pipeline on the commit I opened this PR on are from a rolling build I did on an equivalent branch on dotnet/runtime. This should fix itself in future pushes.

**Comment by jkoritzinsky** [comment]
Job failures are not due to pipeline issues. Can I get another review pass?

**Comment by jakobbotsch** [comment]
Are you sure the jit-cfg failures are unrelated? I don't see the `FileNotFoundException` failures that this PR hit in Sunday's run. E.g.:
```
  Starting:    GC.API.XUnitWrapper (parallel test collections = on [4 threads], stop on fail = off)
    GC\API\GC\Finalize\Finalize.cmd [FAIL]
      Activation Context creation failed. Error: 0x00000002
      Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'C:\h\w\A6E308EA\w\B2070972\e\GC\API\GC\Finalize\Finalize.dll'. The system cannot find the file specified.
      File name: 'C:\h\w\A6E308EA\w\B2070972\e\GC\API\GC\Finalize\Finalize.dll'
```

I [started a run](https://dev.azure.com/dnceng-public/public/_build/results?buildId=591079&view=results) on main so we can double check.

**Comment by jakobbotsch** [comment]
It appears this change is switching the release JIT compilation we use in superpmi-diffs for throughput measurements to have PGO enabled:
image

**Comment by BruceForstall** [comment]
The PGO warning still exists in the triggered diffs run: https://dev.azure.com/dnceng-public/public/_build/results?buildId=591095&view=ms.vss-build-web.run-extensions-tab

**Comment by BruceForstall** [comment]
What is the impact on job / pipeline end-to-end times?

Would the pipeline times benefit from running on more highly parallel build machines (I'm assuming they are not highly parallel today)?

**Comment by BruceForstall** [comment]
There are significant changes that could impact SuperPMI collection and the JIT rolling build. This PR should be tested on those pipelines. That requires pushing the changes to the internal git mirror and manually triggering the relevant pipelines: https://dev.azure.com/dnceng/internal/_build?definitionId=902 and https://dev.azure.com/dnceng/internal/_build?definitionId=977

File: eng/pipelines/coreclr/templates/perf-job.yml
Comment (DrewScoggins) [inline eng/pipelines/coreclr/templates/perf-job.yml:290]: The removal of these steps breaks our mono runs. We expect to have a runnable mono version in .dotnet-mono, and it needs to have the files from the runtime, testhost, and `corerun` combined there. We don't need the build step part of this anymore as we get all of these outputs in the unzipped build.

**Comment by jkoritzinsky** [comment]
@BruceForstall internal runs here:

https://dev.azure.com/dnceng/internal/_build/results?buildId=2397567&view=results
https://dev.azure.com/dnceng/internal/_build/results?buildId=2397571&view=results

**Comment by BruceForstall** [comment]
> @BruceForstall internal runs here:
> 
> https://dev.azure.com/dnceng/internal/_build/results?buildId=2397579&view=results https://dev.azure.com/dnceng/internal/_build/results?buildId=2397571&view=results

Thanks for kicking those off.

Both have problems. The jitrollingbuild one seems to have a permission problem. Maybe that's due to running from a non-main branch? I don't understand it. Maybe dnceng can help.

The superpmi collection pipeline seems to be failing in fundamental ways during the builds.

Is it odd that the jobs have names like "windows-x64 checked" instead of something more descriptive, like "Build product windows x64 checked" or "Build coreclr, libraries, mono windows x64 checked"?

**Comment by jkoritzinsky** [comment]
> > @BruceForstall internal runs here:
> > [dev.azure.com/dnceng/internal/_build/results?buildId=2397579&view=results](https://dev.azure.com/dnceng/internal/_build/results?buildId=2397579&view=results) [dev.azure.com/dnceng/internal/_build/results?buildId=2397571&view=results](https://dev.azure.com/dnceng/internal/_build/results?buildId=2397571&view=results)
> 
> Thanks for kicking those off.
> 
> Both have problems. The jitrollingbuild one seems to have a permission problem. Maybe that's due to running from a non-main branch? I don't understand it. Maybe dnceng can help.

The permissions problem is because of the branch and that I'm not on the list of people with access to the blob storage keys where the results are uploaded.

> The superpmi collection pipeline seems to be failing in fundamental ways during the builds.

Yep I'm looking into it.

> Is it odd that the jobs have names like "windows-x64 checked" instead of something more descriptive, like "Build product windows x64 checked" or "Build coreclr, libraries, mono windows x64 checked"?

The format of the names in the global build job is "platform-arch config CustomSuffix". In the many outerloop pipelines where we are only building one copy of a given os+arch+config combo, I didn't see much of a reason to add a suffix as it adds complexity for the test jobs as I need to specify it there as well and it doesn't add any clarity as to what's being built that can't be trivially found in the pipeline definition.

**Comment by jkoritzinsky** [comment]
Updated superpmi-collect pipeline: https://dev.azure.com/dnceng/internal/_build/results?buildId=2398524&view=results

**Comment by BruceForstall** [comment]
> I didn't see much of a reason to add a suffix as it adds complexity for the test jobs as I need to specify it there as well and it doesn't add any clarity as to what's being built that can't be trivially found in the pipeline definition.

Isn't it useful (IMO, I think it's useful) to have the pipeline name clarify that it is a "build" instead of a "test" run or some other thing?

**Comment by BruceForstall** [comment]
@jkoritzinsky I have a request for your testing:

1. In your branch in the internal mirror, where you are triggering SuperPMI collections, can you please change the JIT-EE GUID? Otherwise, the collections (if they succeed) overwrite the existing "main" collections. I've been having a problem where a recent PR of mine failed superpmi-replay with a C++ exception code when replaying various collections, and I haven't been able to reproduce the problem even when downloading the same build assets used by the job. I'm suspicious that there may be some problem with the collections you created that my PR picked up.
2. After your "private" superpmi collection succeeds, you can use the same JIT-EE GUID in this PR to trigger superpmi replay and asmdiffs pipelines that will use those collections.

**Comment by jkoritzinsky** [comment]
I'll go cancel the jobs I've started and restart them with a custom GUID

**Comment by jkoritzinsky** [comment]
SuperPMI collection run here: https://dev.azure.com/dnceng/internal/_build/results?buildId=2403487&view=results

The diffs and replay jobs I'm re-running now that I've done a collection.

I still need to do some fixes for the wasm-perf legs.

**Comment by BruceForstall** [comment]
I presume you are going to revert your change to jit.h once your testing is done? (It looks like that's causing a formatting job failure, also)

**Review by BruceForstall — APPROVED** [review]
It's basically impossible to review something this massive by sight, but I scanned through various changes, so LGTM.

**Comment by jkoritzinsky** [comment]
> I presume you are going to revert your change to jit.h once your testing is done? (It looks like that's causing a formatting job failure, also)

Yep! Do you want me to keep the JIT-EE GUID change or should I revert that one before I merge this in?

**Comment by BruceForstall** [comment]
> Do you want me to keep the JIT-EE GUID change or should I revert that one before I merge this in?

You should revert it. It was just to avoid trashing our collections when you were testing.

**Comment by jkoritzinsky** [comment]
/ba-g Timeouts in interpreter tests are known

**Review by agocke — APPROVED** [review]
LGTM I think. There might be more work to optimize things, but I think that can come later.

**Comment by akoeplinger** [comment]
@jkoritzinsky thank you so much for doing this work! 🎉

**Comment by kg** [comment]
The most recent commit in https://github.com/dotnet/runtime/pull/99841 didn't run any lanes, is that potentially related to these yml changes?

**Comment by akoeplinger** [comment]
@kg it looks like it's running on AzDO but didn't post back to GitHub: https://dev.azure.com/dnceng-public/public/_build/results?buildId=605726&view=results

**Comment by BruceForstall** [comment]
@jkoritzinsky Looks like libraries-jitstress is broken:

https://dev.azure.com/dnceng-public/public/_build/results?buildId=605612&view=results

```
 Done building Helix work items for scenario no_tiered_compilation. Work item count: 0
D:\a\_work\1\s\src\libraries\sendtohelixhelp.proj(335,5): error : No helix work items, or APKs, or AppBundles found to test
##[error]src\libraries\sendtohelixhelp.proj(335,5): error : No helix work items, or APKs, or AppBundles found to test
```

**Comment by jkoritzinsky** [comment]
@BruceForstall fix for that family of pipelines at https://github.com/dotnet/runtime/pull/99868

**Comment by BruceForstall** [comment]
Looks like `runtime-coreclr outerloop` is broken:

https://dev.azure.com/dnceng-public/public/_build?definitionId=108&_a=summary

```
The 'stages' parameter is not a valid StageList

/eng/pipelines/common/xplat-setup.yml: Could not find /eng/pipelines/common/templates/global-build-job.yml in repository self hosted on https://github.com/ using commit 0935105e91450a1bad02b5b2f83be52bea2bcf59. GitHub reported the error, "Not Found"
 ```

**Comment by jkoritzinsky** [comment]
I've fixed that one as well in #99868 as I had other fixes I had to do there already.

**Pull request merged**
